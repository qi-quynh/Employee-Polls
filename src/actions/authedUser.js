export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_AUTHED_USER = "LOGOUT_AUTHED_USER";

export function setAuthedUser(id) {
  return {
    type: SET_AUTHED_USER,
    id,
  };
}

export function logoutAutherUser() {
  return {
    type: LOGOUT_AUTHED_USER,
  };
}

export function handleLogin(username, password) {
  return (dispatch, getState) => {
    const { users } = getState();
    const user = Object.values(users).find(
      (u) => u.id === username && u.password === password
    );
    if (user) {
      return dispatch(setAuthedUser(user.id));
    } else {
      return dispatch(setAuthedUser(null));
    }
  };
}

export function handleLogout() {
  return (dispatch) => {
    return dispatch(logoutAutherUser);
  };
}
