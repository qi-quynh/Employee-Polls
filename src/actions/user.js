export const GET_USERS = "GET_USERS";
export const UPDATE_ANSWERS_DISPLAY = "UPDATE_ANSWERS_DISPLAY";
export const UPDATE_QUESTIONS_DISPLAY = "UPDATE_QUESTIONS_DISPLAY";

export function receiveUsers(users) {
  return {
    type: GET_USERS,
    users,
  };
}

export function updateAnswersOfUser({ authedUser, qid, answer }) {
  return {
    type: UPDATE_ANSWERS_DISPLAY,
    authedUser,
    qid,
    answer,
  };
}

export function updateQuestionsOfUser(question) {
  return {
    type: UPDATE_QUESTIONS_DISPLAY,
    authedUser: question.author,
    id: question.id,
  };
}
