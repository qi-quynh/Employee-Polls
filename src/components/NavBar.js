import { connect } from "react-redux";
import NavCollapse from "./NavCollapse";
import User from "./User";

const NavBar = (props) => {
  return (
    <div className="navbar-nav">
      <NavCollapse />
      {props.isLogin && <User />}{" "}
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  isLogin: authedUser !== null,
});

export default connect(mapStateToProps)(NavBar);
