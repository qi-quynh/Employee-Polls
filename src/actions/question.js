import { _saveQuestion, _saveQuestionAnswer } from "../utils/_DATA";
import { updateQuestionsOfUser } from "./user";
import { showLoading, hideLoading } from "react-redux-loading-bar";
export const GET_QUESTIONS = "GET_QUESTIONS";
export const ADD_ANSWER = "ADD_ANSWER";
export const ADD_QUESTION = "ADD_QUESTION";

export function getQuestions(questions) {
  return {
    type: GET_QUESTIONS,
    questions,
  };
}

function createAnswer({ authedUser, qid, answer }) {
  return {
    type: ADD_ANSWER,
    authedUser,
    qid,
    answer,
  };
}

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function handleAddAnswer({ authedUser, qid, answer }) {
  return (dispatch) => {
    dispatch(showLoading);
    return _saveQuestionAnswer({
      authedUser,
      qid,
      answer,
    })
      .then(() => {
        dispatch(createAnswer({ authedUser, qid, answer }));
      })
      .then(() => dispatch(hideLoading()));
  };
}

export function handleAddQuestion(question) {
  return (dispatch) => {
    return _saveQuestion(question)
      .then((res) => {
        dispatch(addQuestion(res));
        dispatch(updateQuestionsOfUser(res));
      })
      .catch((e) => console.log("Error from handleSaveQuestion: ", e));
  };
}
