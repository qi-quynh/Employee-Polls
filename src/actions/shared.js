import { _getInitialData } from "../utils/_DATA";
import { receiveUsers } from "./user";
import { getQuestions } from "./question";
import { setAuthedUser } from "./authedUser";

export function handleDataInitial() {
  return (dispatch) => {
    return _getInitialData().then(({ users, questions }) => {
      dispatch(receiveUsers(users));
      dispatch(getQuestions(questions));
    });
  };
}

export function handleLogin(AUTHED_ID) {
  return (dispatch) => {
    dispatch(setAuthedUser(AUTHED_ID));
  };
}
