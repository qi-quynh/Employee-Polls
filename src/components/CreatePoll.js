import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { handleAddQuestion } from "../actions/question";

const CreatePoll = (props) => {
  const navigate = useNavigate();
  const { authedUser, dispatch } = props;

  const data = {
    author: authedUser,
    optionOneText: "",
    optionTwoText: "",
  };

  const [newQuestion, setNewQuestion] = useState(data);
  const [isDisabled, setIsDisabled] = useState(true);

  const checkInput = () => {
    if (newQuestion.optionOneText !== "" && newQuestion.optionTwoText !== "") {
      setIsDisabled(false);
    }
  };

  useEffect(() => {
    checkInput();
  }, [newQuestion]);

  const handleSubmit = (e) => {
    e.preventDefault();

    dispatch(handleAddQuestion(newQuestion)).then(() => {
      setIsDisabled(true);
      setNewQuestion(data);
      navigate("/");
    });
  };

  const onValueChange = (e) => {
    const { name, value } = e.target;
    setNewQuestion({ ...newQuestion, [name]: value });
  };

  return (
    <div className="create">
      <h1 className="create-header">Create Poll</h1>
      <form className="create-form" onSubmit={handleSubmit}>
        <span className="create-display">Would you rather</span> <br />
        <input
          name="optionOneText"
          className="input"
          type="text"
          size="70"
          value={newQuestion.optionOneText}
          onChange={onValueChange}
        />
        or <br />
        <input
          name="optionTwoText"
          className="input"
          size="70"
          type="text"
          onChange={onValueChange}
          value={newQuestion.optionTwoText}
        />
        ?
        <br />
        <button disabled={isDisabled} className="button-create">
          Create
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => {
  return { authedUser };
};

export default connect(mapStateToProps)(CreatePoll);
