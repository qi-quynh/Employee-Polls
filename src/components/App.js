import { Fragment, useEffect } from "react";
import { connect } from "react-redux";
import { handleDataInitial } from "../actions/shared";
import NavBar from "./NavBar";
import "../index.css";
import { Route, Routes } from "react-router-dom";
import Home from "./Home";
import CreatePoll from "./CreatePoll";
import LeaderBoard from "./LeaderBoard";
import PollPage from "./PollPage";
import Error404 from "./Error404";
import Login from "./login";

const App = (props) => {
  const { isLogin } = props;
  useEffect(() => {
    props.dispatch(handleDataInitial());
  }, [isLogin]);
  return (
    <Fragment>
      <div className="container">
        <NavBar />
        {isLogin === true ? (
          <Routes>
            <Route exact path="/" element={<Home />} />
            <Route path="/add" element={<CreatePoll />} />
            <Route path="/questions/:question_id" element={<PollPage />} />
            <Route path="/leaderboard" element={<LeaderBoard />} />
            <Route path="*" element={<Error404 />} />
          </Routes>
        ) : (
          <Login />
        )}
      </div>
    </Fragment>
  );
};
const mapStateToProps = ({ authedUser }) => ({
  isLogin: authedUser !== null,
});
export default connect(mapStateToProps)(App);
