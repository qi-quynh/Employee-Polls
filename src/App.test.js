import { render, screen, fireEvent } from "@testing-library/react";
import App from "./components/App";
import NavBar from "./components/NavBar";
import "@testing-library/jest-dom/extend-expect";
import { MemoryRouter, BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "./store";
import { handleDataInitial } from "./actions/shared";
import Login from "./components/login";
import CreatePoll from "./components/CreatePoll";
const {
  _saveQuestionAnswer,
  _saveQuestion,
  generateUID,
} = require("../src/utils/_DATA");
describe("Navbar", () => {
  test("render correct links", async () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <NavBar />
        </Provider>
      </MemoryRouter>
    );

    const home = screen.getByText(/home/i);
    const leaderboard = screen.getByText(/leaderboard/i);
    const newPoll = screen.getByText(/new poll/i);
    expect(home).toBeInTheDocument();
    expect(newPoll).toBeInTheDocument();
    expect(leaderboard).toBeInTheDocument();
  });
});

describe("Navbar", () => {
  it("should render the component", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });
});

describe("Save question with success testing", () => {
  it("should return success if the optionOne, optionTwo, author are correct", async () => {
    const question = {
      optionOneText: "Would you like to go to the cinema",
      optionTwoText: "Would you like to go to dinner",
      author: "sarahedo",
    };
    var res = await _saveQuestion(question);

    expect(res.optionOne.text).toEqual("Would you like to go to the cinema");
    expect(res.optionTwo.text).toEqual("Would you like to go to dinner");
    expect(res.author).toEqual("sarahedo");
  });
});
describe("Save question with errors testing", () => {
  it("should return error if optionOneText is null", async () => {
    const question = {
      optionTwoText: "Would you like coffee",
      author: "mtsamis",
    };
    await expect(_saveQuestion(question)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });

  it("should return error if optionTwoText is null", async () => {
    const question = {
      optionOneText: "Would you like juice",
      author: "mtsamis",
    };
    await expect(_saveQuestion(question)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });

  it("should return error if author is null", async () => {
    const question = {
      optionTwoText: "Would you like coffee",
      optionOneText: "Would you like juice",
    };
    await expect(_saveQuestion(question)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });
});
describe("Create new poll", () => {
  test("button will be disabled if both 1st and 2nd cell are entered", async () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <CreatePoll />
        </Provider>
      </MemoryRouter>
    );

    const oEnter = screen.getByTestId("text-one-test");
    const tEnter = screen.getByTestId("text-two-test");
    const btn = screen.getByTestId("submit-test");

    expect(btn).toHaveAttribute("disabled");

    fireEvent.change(oEnter, { target: { value: "Would you like coffee" } });
    fireEvent.change(tEnter, { target: { value: "Would you like juice" } });

    expect(btn).not.toHaveAttribute("disabled");
  });
});
describe("UID", () => {
  test("should return a unique question id", () => {
    const question1 = generateUID();
    const question2 = generateUID();
    expect(question1 !== question2).toBe(true);
  });
});

describe("_saveQuestionAnswer", () => {
  it("should return true for correct parameters", async () => {
    const result = await _saveQuestionAnswer({
      authedUser: "tylermcginnis",
      qid: "xj352vofupe1dqz9emx13r",
      answer: "optionTwo",
    });

    expect(result).toBeTruthy();
  });

  it("should return error for false parameters", async () => {
    const result = await _saveQuestionAnswer({
      authedUser: "tylermcginnis",
      qid: undefined,
      answer: "optionTwo",
    }).catch((e) => e);

    expect(result).toBe("Please provide authedUser, qid, and answer");
  });
});
