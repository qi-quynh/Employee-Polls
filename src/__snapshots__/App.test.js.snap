// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="navbar-nav"
        >
          <ul>
            <li
              style="display: inline;"
            >
              <a
                class="item non"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              style="display: inline;"
            >
              <a
                class="item non"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li
              style="display: inline;"
            >
              <a
                class="item non"
                href="/add"
              >
                New Poll
              </a>
            </li>
          </ul>
           
        </div>
        <div>
          <h1
            class="signIn-heading"
          >
            LOGIN
          </h1>
          <div
            class="signIn"
          >
            <h1>
              Login
            </h1>
            <select
              name="signIn-select"
            >
              <option>
                Suggestion Member
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="navbar-nav"
      >
        <ul>
          <li
            style="display: inline;"
          >
            <a
              class="item non"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            style="display: inline;"
          >
            <a
              class="item non"
              href="/leaderboard"
            >
              Leaderboard
            </a>
          </li>
          <li
            style="display: inline;"
          >
            <a
              class="item non"
              href="/add"
            >
              New Poll
            </a>
          </li>
        </ul>
         
      </div>
      <div>
        <h1
          class="signIn-heading"
        >
          LOGIN
        </h1>
        <div
          class="signIn"
        >
          <h1>
            Login
          </h1>
          <select
            name="signIn-select"
          >
            <option>
              Suggestion Member
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
